### DISCUSSION TITLE
Stubs for static API cause false positives for MethodReturnCheck

### DISCUSSION THREAD

**Discussion by serianox** [spotbugs/spotbugs#2781] [Q&A]
I'm running analysis on a project for which some external APIs are provided to us as stubs, such as one below.

```java
public static int doSomething(int arg1, int arg2) {
    // TODO Auto-generated method stub
    return 0;
}
```

However, it seems that in the case of a static method, those gets analyzed by Spotbugs and marked as having no side effects. Which means that calling any of those methods results in a RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT getting generated.

I understand that I can annotate using @CheckReturnValue, but I don't have direct control over the API getting analyzed. And I'm looking for ways other than @SuppresFBWarnings or deactivating MethodReturnCheck entirely.

I've also seen in the code that Spotbugs knows the patterns for auto-generated stubs, but those are only used for instance methods, not static methods.

Is there any way to indicate to Spotbugs that those are stubs?

**Comment by ThrawnCA** [comment]
You should be able to use a [filter file](https://spotbugs.readthedocs.io/en/stable/filter.html) to ignore the methods involved.

**Reply by serianox** [reply]
Thanks for your input!

I get the following output from Spotbugs (I apologize for the redacted parts):

```xml
  <BugInstance type="RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT" priority="2" rank="17" abbrev="RV" category="STYLE" instanceHash="xxx" instanceOccurrenceNum="0" instanceOccurrenceMax="0" cweid="440">
    <ShortMessage>Return value of method without side effect is ignored</ShortMessage>
    <LongMessage>Return value of callee(short, short, short) ignored, but method has no side effect</LongMessage>
    <Class classname="xxx" primary="true">
      <SourceLine classname="xxx" start="336" end="3685" sourcefile="xxx" sourcepath="xxx">
        <Message>xxx</Message>
      </SourceLine>
      <Message>xxx</Message>
    </Class>
    <Method classname="xxx" name="caller" signature="xxx" isStatic="false" primary="true">
      <SourceLine classname="xxx" start="1431" end="1847" startBytecode="0" endBytecode="2649" sourcefile="xxx" sourcepath="xxx"/>
      <Message>xxx</Message>
    </Method>
    <Method classname="xxx" name="callee" signature="(SSS)S" isStatic="true" role="METHOD_CALLED">
      <SourceLine classname="xxx" start="683" end="683" startBytecode="0" endBytecode="25" sourcefile="xxx" sourcepath="xxx"/>
      <Message>xxx</Message>
    </Method>
    <SourceLine classname="xxx" primary="true" start="1510" end="1510" startBytecode="338" endBytecode="338" sourcefile="xxx" sourcepath="xxx">
      <Message>xxx</Message>
    </SourceLine>
```

With two _Method_ node, the first one being the caller and the second one the callee.

I assume from the documentation that I can use the following filter and specify either the caller or the callee in the _Match_ node. If I specify the caller, I get all call suppressed. But if I specify the callee instead (to suppress all errors from calling the callee only), no false positive gets suppressed.

Is it supposed to work this way?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<FindBugsFilter xmlns="https://github.com/spotbugs/filter/3.0.0"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                xsi:schemaLocation="https://github.com/spotbugs/filter/3.0.0 https://raw.githubusercontent.com/spotbugs/spotbugs/3.1.0/spotbugs/etc/findbugsfilter.xsd">
    <Match>
        <Method name="callee" />
        <Bug pattern="RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT" />
    </Match>
</FindBugsFilter>
```

**Reply by serianox** [reply]
update: I'm currently using _@SuppressFBWarnings("RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT")_ which is an okay-ish workaround. Still, I need to annotate the caller, so I may miss true positives.

One of the better way could be to reissue the mentioned API annotated with _@CanIgnoreReturnValue_, but I wish not to re-annotate external API. 😇

**Reply by ThrawnCA** [reply]
The caller is the point where SpotBugs thinks the behaviour (throwing away the return value) is questionable. The callee isn't doing anything that raises flags.
